validator/sawtooth_validator/gossip/gossip.py

```diff
@@ -228,7 +228,14 @@ def register_peer(self, connection_id, endpoint):
                 peer
         """
         with self._lock:
+            # If an endpoint is already peered with different connection id
+            # do no allow another peering request.
             if len(self._peers) < self._maximum_peer_connectivity:
+                if endpoint in self._peers.values():
+                    raise PeeringException(
```

Comment:
What are the consequences of raising this exception? Should this be idempotent, and return?

Reply:
I had similar question, discussion is still open in rocketchat channel #sawtooth-core-dev. By raising exception here current connection would be closed and a new connection would be attempted as per the retry logic configured in gossip part. If in case we plan to return instead of exception, let's consider two cases and see how it impacts

Case 1 (connection initiated by current peer): The connection initiator part of code should not consider peering request to be successful.  It may however need a connection retry mechanism in place.

Case 2 (connection initiated by other peer): Other peer should not be replied back as if peering request is accepted.

Both of which means return value should be a boolean so that a decision can be made at caller end. But I guess as per coding convention it makes sense to introduce another exception if not this, instead of changing function signature to return boolean.

I am open for comments, please feel free to suggest best practice